lib/activerecord/delay_touching/state.rb

```diff
@@ -49,6 +49,16 @@ def clear_records
         @already_updated_records.clear
       end
 
+      # If we don't do this then an infinite loop is possible due to how 
```

Comment:
FWIW, if you are going to comment something, I'd mention that ActiveRecord lazily syncs transaction state after rollback, which my change in-memory state of key objects, which requires that the hash be rekeyed.

Set#subtract and ActiveRecord::Core#== work as expected (and aren't worth mentioning, IMO); it's really the lazy updating on rollbacks that's the kicker.